app/main.py

```diff
@@ -1,213 +1,20 @@
-
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
-
-        if "protection" in arthur["potion"]["effect"]:
-            arthur["protection"] += arthur["potion"]["effect"]["protection"]
-
-        if "hp" in arthur["potion"]["effect"]:
-            arthur["hp"] += arthur["potion"]["effect"]["hp"]
-
-    # mordred
-    mordred = knightsConfig["mordred"]
-
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
-
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
-
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
-
-        if "protection" in mordred["potion"]["effect"]:
-            mordred["protection"] += mordred["potion"]["effect"]["protection"]
-
-        if "hp" in mordred["potion"]["effect"]:
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
-
-    # red_knight
-    red_knight = knightsConfig["red_knight"]
-
-    # apply armour
-    red_knight["protection"] = 0
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
-
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
-
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
-
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
-
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
-
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
-
-    # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
+from app.action import fight
+from app.knights import KNIGHTS
+from app.preparation import preparations
 
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
 
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
+def battle(knights: dict) -> dict:
+    # prepare all knights
+    knights_dict = preparations(knights)
+    # knights are fighting here
+    fight(knights_dict["lancelot"], knights_dict["mordred"])
```

Comment:
What if I add a new pair of knights? You will be need to add them here as well, so your code still is not reusable

Reply:
this question is not correct, because in any case, this pair should be written in the code, who will fight with whom. and if you don't even add more knights, but simply change their places, even then you have to rewrite the condition. So make code that can be reused, only with addition input function where you can manually enter the name of the knight and his opponent.

Reply:
But the knights number are not specific, you have an example in `KNIGHTS` variable. Try to add 2 new knights and rerun your code
![image](https://github.com/mate-academy/py-knights-fighting/assets/73850362/e8c6a551-d16b-49ac-a7a2-88e092723800)

Reply:
nothing will change because they are just extra knights. They will not affect the solution of the main condition of the task. What do you want me to do neither classes nor functions can solve this problem. because you still need to know who is fighting whom. In condition 1, it fights with 3m. 2 out of 4. If I add 5 and 6 knights, where is the challenge that they have to fight each other instead of 1 out of 5, 2 out of 6? That's why I say - this is not a correct question.